p2p/peer_set.go

```diff
@@ -120,24 +120,22 @@ func (ps *PeerSet) Remove(peer Peer) bool {
 	}
 
 	index := item.index
-	// Create a new copy of the list but with one less item.
-	// (we must copy because we'll be mutating the list).
-	newList := make([]Peer, len(ps.list)-1)
-	copy(newList, ps.list)
+	lastPeer := ps.list[len(ps.list)-1]
+	ps.list[len(ps.list)-1] = nil // nil the last entry of the slice to shorten, so it isn't reachable & can be GC'd.
+
 	// If it's the last peer, that's an easy special case.
 	if index == len(ps.list)-1 {
-		ps.list = newList
+		ps.list = ps.list[:len(ps.list)-1]
 		delete(ps.lookup, peer.ID())
 		return true
 	}
 
 	// Replace the popped item with the last item in the old list.
-	lastPeer := ps.list[len(ps.list)-1]
 	lastPeerKey := lastPeer.ID()
 	lastPeerItem := ps.lookup[lastPeerKey]
-	newList[index] = lastPeer
+	ps.list[index] = lastPeer
+	ps.list = ps.list[:len(ps.list)-1]
```

Comment:
Could you move this line to where it was before? The two changes that happen in both cases is to reduce `ps.list` by one element and removing the key from `ps.lookup`.